File: Homework2/Homework2/Homework2/joinDistribution.js
class JoinDistribution {
    constructor(matrix, numCols, numRows) {
        this.matrix = matrix;
        this.numCols = numCols;
        this.numRows = numRows;
    }

    printDistribution() {
        const evalDist = this.evalDistribution();

        for (const [key, value] of Object.entries(evalDist)) {
            console.log(`Key = ${key}, Value = ${value}`);
        }
        console.log("\n");
    }

    evalDistribution() {
        const jointDistribution = {};
        const variables = ["Ambitious (0-5)", "Programming Languages"]; // Puoi inserire più variabili

        const varColumns = new Array(variables.length).fill(0);
        for (let i = 0; i < variables.length; i++) {
            for (let j = 0; j < this.numCols; j++) {
                if (this.matrix[0][j] === variables[i]) {
                    varColumns[i] = j;
                    break;
                }
            }
        }

        const valuesMatrix = new Array(varColumns.length);
        let jointValue;

        for (let i = 1; i < this.numRows; i++) {
            valuesMatrix[0] = this.matrix[i][varColumns[0]].toLowerCase().replace(/["\s,]/g, '').split(',');
            for (let k = 1; k < varColumns.length; k++) {
                valuesMatrix[k] = this.matrix[i][varColumns[k]].toLowerCase().replace(/["\s,]/g, '').split(',');
            }

            const combinations = this.cartesianProduct(valuesMatrix);
            for (const combination of combinations) {
                jointValue = combination.join(", ");

                if (!jointDistribution.hasOwnProperty(jointValue)) jointDistribution[jointValue] = 1;
                else jointDistribution[jointValue]++;
            }
        }

        // Ordina e restituisci il risultato
        const sortedJointDistribution = Object.fromEntries(
            Object.entries(jointDistribution).sort((a, b) => b[1] - a[1])
        );

        return sortedJointDistribution;
    }

    cartesianProduct(items) {
        const currentItem = new Array(items.length);
        return this.go(items, currentItem, 0);
    }

    *go(items, currentItem, index) {
        if (index === items.length) {
            yield [...currentItem];
        } else {
            for (const item of items[index]) {
                currentItem[index] = item.toString();
                for (const j of this.go(items, currentItem, index + 1)) {
                    yield [...j];
                }
            }
        }
    }
}

// Esempio di utilizzo
const matrix = [/* Inserire la matrice qui */];
const numCols = /* Numero di colonne */;
const numRows = /* Numero di righe */;

const distribution = new JoinDistribution(matrix, numCols, numRows);
distribution.printDistribution();

File: Homework2/Homework2/Homework2/JoinDistribution.cs
// Distributions.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.Eventing.Reader;
using System.Linq;

namespace Homework2
{
    public partial class JoinDistribution
    {
        string[][] matrix;
        int numCols;
        int numRows;

        public JoinDistribution(string[][] matrix,int numCols,int numRows)
        {
            this.matrix = matrix;
            this.numCols = numCols;
            this.numRows = numRows;
        }

        public void printDistribution()
        {
            Dictionary<string, int> evalDist = this.evalDistribution();

            foreach (var kvp in evalDist)
            {
                Console.WriteLine("Key = {0}, Value = {1}", kvp.Key, kvp.Value);
            }
            Console.WriteLine("\n");
        }

        // works for multivaried and univaried
        private Dictionary<string, int> evalDistribution()
        {
            Dictionary<string,int> jointDistribution = new Dictionary<string, int>();
            string[] variables = { "Ambitious (0-5)", "Programming Languages" }; //can insert variables multiple than 2  

            int[] varColumns = new int[variables.Length];
            for (int i = 0; i < variables.Length; i++)
            {
                for (int j = 0; j < numCols; j++)
                {
                    if (matrix[0][j] == variables[i])
                    {
                        varColumns[i] = j;
                        break;
                    }
                }
            }

            string[][] valuesMatrix = new string[varColumns.Length][];
            string jointValue;

            for (int i = 1; i < numRows; i++)
            {
                valuesMatrix[0] = matrix[i][varColumns[0]].ToLower().Trim('"').Trim(' ').Trim(',').Split(',');
                for (int k = 1; k < varColumns.Length; k++)
                {
                    valuesMatrix[k] = matrix[i][varColumns[k]].ToLower().Trim('"').Trim(' ').Trim(',').Split(',');
                }

                var combinations = CartesianProduct(valuesMatrix);
                foreach (var combination in combinations)
                {
                    jointValue = String.Join(", ", combination);

                    if (!jointDistribution.ContainsKey(jointValue)) jointDistribution[jointValue] = 1;
                    else jointDistribution[jointValue]++;
                }

            }

            return jointDistribution.OrderByDescending(f => f.Value).ToDictionary(f => f.Key, f => f.Value);
        }
        private static IEnumerable<string[]> CartesianProduct(string[][] items)
        {
            string[] currentItem = new string[items.Length];
            return Go(items, currentItem, 0);
        }

        private static IEnumerable<string[]> Go(string[][] items, string[] currentItem, int index)
        {
            if (index == items.Length)
            {
                yield return (string[])currentItem.Clone();
            }
            else
            {
                foreach (string item in items[index])
                {
                    currentItem[index] = item.ToString();
                    foreach (string[] j in Go(items, currentItem, index + 1))
                    {
                        yield return (string[])j.Clone();
                    }
                }
            }
        }
    }
}